node/src/chain/fsm.rs

```diff
@@ -150,30 +149,32 @@ impl<N: Network, DB: database::DB, VM: vm::VMExecution> SimpleFSM<N, DB, VM> {
     /// A block event could originate from either local consensus execution, a
     /// wire Block message (topics::Block), or a wire Quorum message
     /// (topics::Quorum).
+    ///
+    /// If the block is accepted, it return the block itself
     pub async fn on_block_event(
         &mut self,
-        blk: &Block,
+        blk: Block,
         metadata: Option<Metadata>,
-    ) -> anyhow::Result<()> {
+    ) -> anyhow::Result<Option<Block>> {
+        let block_hash = blk.header().hash;
+
         // Filter out blocks that have already been marked as
         // blacklisted upon successful fallback execution.
-        if self
-            .blacklisted_blocks
-            .read()
-            .await
-            .contains(&blk.header().hash)
-        {
+        if self.blacklisted_blocks.read().await.contains(&block_hash) {
             info!(
                 event = "block discarded",
                 reason = "blacklisted",
                 hash = to_str(&blk.header().hash),
                 height = blk.header().height,
                 iter = blk.header().iteration,
             );
-            return Ok(());
+            // block discarded, should we clean up attestation cache (if any)?
```

Comment:
I believe clearing up the cache is not necessary. Blocks are blacklisted only when they are reverted (which means they were already full blocks, which means their attestation was not in the cache, IIUC)

Reply:
I moved the clear of the cache inside the att_if_needed method

Reply:
we can probably remove the second part of this comment